Throw if the provided LaunchSettings use the same url for applicationURL and DOTNET_DASHBOARD_OTLP_ENDPOINT_URL

ProsserCJ opened the pull request:
Fixes #3044

Add a check to DistributedApplicationBuilder.Build(), which throws if the provided LaunchSettings use the same url for applicationURL and DOTNET_DASHBOARD_OTLP_ENDPOINT_URL.

Includes corresponding test.
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/3419)

[edited by @danmoseley  to add bug link]

ProsserCJ committed:
Fixes #3044

Add a check to DistributedApplicationBuilder.Build(), which throws if the provided LaunchSettings use the same url for applicationURL and DOTNET_DASHBOARD_OTLP_ENDPOINT_URL.

Includes corresponding test

davidfowl commented:
I think this should move this  logic here: https://github.com/dotnet/aspire/blob/516e8892e2a098e9b2b8b3926aa007b44c89152a/src/Aspire.Hosting/Dashboard/DashboardOptions.cs#L41

ProsserCJ committed:
Merge remote-tracking branch 'origin/main' into issue-3044

ProsserCJ committed:
Fixes #3044

Remove previous fix.

Add check to Aspire.Hosting.ValidateDashboardOptions to add error if DashboardUrl == OtlpEndpointUrl.

Add DashboardOptionsValidatorTests with tests for the three error cases.

ProsserCJ commented:
@davidfowl nice, looks much cleaner. Added new commit

davidfowl commented on tests/Aspire.Hosting.Tests/Dashboard/DashboardOptionsValidatorTests.cs:
```diff
@@ -0,0 +1,64 @@
+    [Fact]
+    public void ValidateFailWhenDashboardUrlIsNullOrEmpty()
+    {        
+        foreach(string? url in new List<string?> { null, string.Empty })
```
Can you use a theory + inlinedata instead of using a loop in the tests?

davidfowl commented on src/Aspire.Hosting/Dashboard/DashboardOptions.cs:
```diff
@@ -48,6 +48,11 @@ public ValidateOptionsResult Validate(string? name, DashboardOptions options)
             builder.AddError("Failed to configure dashboard resource because DOTNET_DASHBOARD_OTLP_ENDPOINT_URL environment variable was not set.");
         }
 
+        if (options.DashboardUrl == options.OtlpEndpointUrl)
```
case insensitive comparison?

ProsserCJ committed:
Iterate on #3044

Use case-insensitive string compare. Use Theories with inline data instead of Facts

davidfowl commented:
There are test failures

ProsserCJ committed:
#3044: Fix broken tests

Some tests initialize both ASPNETCORE_URLS and DOTNET_DASHBOARD_OTLP_ENDPOINT_URL to the same url, which is no longer permitted.

ProsserCJ commented:
@davidfowl tests fixed. Anything else?

davidfowl commented:
/ap run

davidfowl commented:
/azp run

davidfowl commented:
@ProsserCJ very sorry this PR went stale, if you want to re-implement let me know. Otherwise feel free to close the issue. Not many people have run into this.
